{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8d80a72d",
   "metadata": {},
   "source": [
    "# ~ Beagle 5.4 Homozygous Refererence Imputation Accuracy Experiment ~\n",
    "# Part 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eb6d489",
   "metadata": {},
   "source": [
    "### - This small experiment is meant to give an indication of how well the imputation program Beagle 5.4 can impute homozygous reference regions by only using variant calls for heterozygous and homozygous variants as input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "788acd19",
   "metadata": {},
   "source": [
    "### |====================================================================================|"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05f94704",
   "metadata": {},
   "source": [
    "# General Steps:\n",
    "## 1. Create a complete VCF format dataframe of chromosome 1, including all homozygous references extracted from reference assembly\n",
    "##    - This will function as the golden reference that the imputed chromosome region will be compared against\n",
    "## 2. Impute a region of chromosome 1, with the imputation only using heterozygous and homozygous alt variants as input\n",
    "## 3. Calculate concordance between imputed chr1 region and actual chr1 region"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa85fc2",
   "metadata": {},
   "source": [
    "### Libraries and Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bd25323d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import multiprocessing\n",
    "import time\n",
    "import re\n",
    "import wget\n",
    "import os \n",
    "import shutil\n",
    "import zipfile\n",
    "import json\n",
    "import subprocess\n",
    "import psutil\n",
    "import sys\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "41a0232d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code From 1_preprocess_CGI_v2.py\n",
    "\n",
    "def read_CGI(cgi_path):\n",
    "\n",
    "    tr0 = time.time()\n",
    "    print(f\"\\tReading file...\")\n",
    "\n",
    "    col_line = None\n",
    "    with open(cgi_path, 'r') as f:\n",
    "        for line in f:\n",
    "            if '>' in line or 'locus' in line:\n",
    "                col_line = col_line\n",
    "                break\n",
    "\n",
    "    cols = line.strip('\\n').rsplit('\\t')\n",
    "    col_types = {col:str for col in cols}\n",
    "\n",
    "    chunks = pd.read_csv(cgi_path, sep='\\t', names = cols, header = None, comment = '#', dtype=col_types, chunksize=100000)\n",
    "    df = pd.concat(chunks)\n",
    "    df = df.iloc[1:]\n",
    "\n",
    "    tr1 = round(time.time()-tr0,2)\n",
    "    print(f\"\\tFinished reading, took {tr1}s\")\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_header_build(file_path):\n",
    "    \n",
    "    with open(file_path,\"r\") as f:\n",
    "        header_info = []\n",
    "        while True:\n",
    "            line = f.readline()\n",
    "            if line.startswith(\"#\"):\n",
    "                line = line.replace('\\t', \" \")\n",
    "                line = line.replace('\\n', \"\")\n",
    "                header_info.append(line[1:])\n",
    "            else:\n",
    "                break\n",
    "                \n",
    "        build = None\n",
    "                \n",
    "        for x in header_info:\n",
    "            if 'GENOME_REFERENCE' in x:\n",
    "                if '36' in x:\n",
    "                    print('Build 36')\n",
    "                    build = 36\n",
    "                elif '37' in x:\n",
    "                    print('Build 37')\n",
    "                    build = 37\n",
    "                elif '38' in x:\n",
    "                    print(\"Build 38\")\n",
    "                    build = 38\n",
    "                else:\n",
    "                    build = 'Unknown Build'\n",
    "                    print(\"Unknown Build\")\n",
    "\n",
    "        return build,\n",
    "\n",
    "def get_col_name(df, names):\n",
    "    #returns the column string that has a matching search term in the list names\n",
    "\n",
    "    cols = df.columns\n",
    "    try:\n",
    "        col_name = [col for col in cols if col in names][0]\n",
    "    except Exception as e:\n",
    "        return None\n",
    "    return col_name\n",
    "\n",
    "\n",
    "### Custom Filter script -> Leaves in homozygous reference region calls ###\n",
    "\n",
    "def filter_chr(df):\n",
    "    var_types = df['varType'].unique().tolist()\n",
    "    var_types.remove('snp')\n",
    "    var_types.remove('ref')\n",
    "    remove_types = var_types\n",
    "    \n",
    "    #df = df[df['reference'].str.contains('=') == False] #Genotype is same as reference\n",
    "    df = df[df['varType'].isin(remove_types) == False] #Exclude variant types that are not snp or reference equivalent\n",
    "    df = df[df['reference'].str.len() == 1] #remove genotypes spanning multiple positions in a single line\n",
    "    df = df[df['alleleSeq'].str.len() == 1] #remove genotypes spanning multiple positions in a single line\n",
    "    \n",
    "    \n",
    "    #Remove low quality calls\n",
    "    accepted_calls = ['PASS','VQHIGH',np.nan]\n",
    "    call_rating_col = get_col_name(df, ['varQuality','varFilter']) #call rating column has a different name in different CGI files\n",
    "    if call_rating_col:\n",
    "        df = df[df[call_rating_col].isin(accepted_calls) == True]\n",
    "        \n",
    "    \n",
    "    #Remove Mitochondrial SNPs\n",
    "    df = df[df['chromosome'].str.contains('M') == False]\n",
    "    \n",
    "    \n",
    "#     #Remove variation loci with single occurence: all should have a partner allele for determination of homozygous / heterozygous for alt\n",
    "    \n",
    "#     df_variants = df[df['reference']!='=']\n",
    "#     are_duplicates = df_variants.duplicated('>locus', keep=False)\n",
    "#     df_variants = df_variants[are_duplicates]\n",
    "    \n",
    "#     df_ref = df[df['reference'] == '=']\n",
    "    \n",
    "#     df = pd.concat([df_variants,df_ref])\n",
    "#     df['begin'] = df['begin'].astype(int)\n",
    "#     df['>locus'] = df['>locus'].astype(int)\n",
    "#     df.sort_values(by=['begin','>locus'], inplace = True)\n",
    "    \n",
    "#     #Remove loci with single occurence (all loci should have a partner, those with single occurence had partner removed in one of the previous steps)\n",
    "#     are_duplicates = df.duplicated('>locus', keep=False)\n",
    "#     df = df[are_duplicates]\n",
    "    \n",
    "    #If rsid missing from external reference OR multiple rsids are present, replace with \".\"\n",
    "    df['xRef'] = df['xRef'].fillna(\".\")\n",
    "    df.loc[df['xRef'].str.contains(';'), 'xRef'] = '.'\n",
    "    \n",
    "    #Keep only rsid\n",
    "    df['xRef'] = df['xRef'].str.split(':').str[1]\n",
    "    df['xRef'] = df['xRef'].fillna(\".\")\n",
    "    \n",
    "    \n",
    "#     df.reset_index(drop=True, inplace=True)\n",
    "    return df\n",
    "\n",
    "def build_vcf_df(cgi_df):\n",
    "    \n",
    "    df = cgi_df\n",
    "    \n",
    "    #Rename columns\n",
    "    df.rename(columns = {'xRef':'ID', \n",
    "                         'end':'POS', \n",
    "                         'reference':'REF', \n",
    "                         'alleleSeq':'ALT', \n",
    "                         'chromosome':'CHROM'}, inplace = True)\n",
    "    \n",
    "    #Create QUAL, FILTER, INFO, FORMAT and NA00001 col \n",
    "    df['QUAL'] = '.'\n",
    "    df['FILTER'] = '.'\n",
    "    df['INFO'] = '.'\n",
    "    df['FORMAT'] = 'GT'\n",
    "    #df['NA00001'] = None\n",
    "    \n",
    "    #Remove loci with single occurence (all loci should have a partner, those with single occurence had partner removed in one of the previous steps)\n",
    "    are_duplicates = df.duplicated('>locus', keep=False)\n",
    "    df = df[are_duplicates]\n",
    "\n",
    "    #Populate NA00001 with heterozygous/homozygous for alt allele \n",
    "    df['orig_ref_alt'] = df['REF'] + df['ALT']\n",
    "  \n",
    "    df_alt_count = df.groupby('>locus')['orig_ref_alt'].agg(list).reset_index()\n",
    "\n",
    "    df_alt_count['ref'] = df_alt_count['orig_ref_alt'].str.get(0)\n",
    "    df_alt_count['alt'] = df_alt_count['orig_ref_alt'].str.get(1)\n",
    "    df_alt_count['alt1'] = df_alt_count['alt'].str[1]\n",
    "    df_alt_count['alt2'] = df_alt_count['ref'].str[1]\n",
    "    df_alt_count['ref'] = df_alt_count['ref'].str[0]\n",
    "\n",
    "    df_alt_count['NA00001'] = np.where(df_alt_count['alt1'] != df_alt_count['alt2'], '0/1', '1/1')\n",
    "    df_alt_count['NA00001'] = np.where((df_alt_count['alt1'] == df_alt_count['alt2']) & (df_alt_count['alt1'] == df_alt_count['ref']), '0/0', df_alt_count['NA00001'])\n",
    "\n",
    "    df_alt_count = df_alt_count[['>locus', 'NA00001']]\n",
    "\n",
    "    df = df.merge(df_alt_count, on='>locus', how='left')\n",
    "\n",
    "    df['2nd_sort_col'] = np.where(df['REF'] != df['ALT'], 2, 1)\n",
    "    #df['begin'] = df['begin'].astype(int)\n",
    "    df = df.sort_values(by=['>locus', '2nd_sort_col'], ascending = False) #Very important for duplicates filter in following step\n",
    "    \n",
    "    \n",
    "    #Replace alt alleles that are actually reference with \".\"\n",
    "    df['ALT'] = np.where(df['REF'] == '=', \".\", df['ALT'])\n",
    "\n",
    "    are_duplicates = df.duplicated('>locus', keep='first')\n",
    "    inverted = ~are_duplicates\n",
    "    df = df[inverted]\n",
    "    \n",
    "    #Remove non-relevant columns\n",
    "    df = df[['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'NA00001']]\n",
    "    df['CHROM'] = df['CHROM'].str[3:] #Remove 'chr' prefix\n",
    "    \n",
    "    #Fill missing rsids\n",
    "    df['ID'] = df['ID'].fillna(\".\")\n",
    "    df['POS'] = df['POS'].astype(int)\n",
    "    df = df.sort_values(by=['POS'], ascending = True)\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def save_file(df, input_file_path, output_path):\n",
    "    #Save processed file as a 23andMe-like .txt file\n",
    "\n",
    "    build = get_header_build(input_file_path)\n",
    "    new_header = ['fileformat=VCFv4.2',\n",
    "                  'source=custom vcf transform script',\n",
    "                  'reference={}'.format(build), \n",
    "                  'FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype>\"']\n",
    "                                  \n",
    "    columns = df.columns.tolist()\n",
    "\n",
    "    with open(output_path, 'w') as f:\n",
    "        #Write Header\n",
    "        for item in new_header: \n",
    "            f.write('##' + item +'\\n')\n",
    "\n",
    "        #Write Columns\n",
    "        f.write('#')\n",
    "        for col in columns:\n",
    "            if col!= 'NA00001':\n",
    "                f.write(col+'\\t')\n",
    "            else:\n",
    "                f.write(col)\n",
    "                f.write('\\n')\n",
    "\n",
    "    df.to_csv(r'{}'.format(output_path), header=False, index=None, sep='\\t', mode='a')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8999bea1",
   "metadata": {},
   "source": [
    "# Load in File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "847caa98",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Complete Genomics raw input file\n",
    "input_f = '/Users/jerenolsen/Desktop/Genomics_Pipeline_Versions/Genomics_Pipeline/input_genomes/hu1AF4ED_CGI_RawData.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "42247041",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tReading file...\n",
      "\tFinished reading, took 29.59s\n"
     ]
    }
   ],
   "source": [
    "df = read_CGI(input_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0606c07",
   "metadata": {},
   "source": [
    "### Comparing count of homozygous reference calls to variant calls in file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5c0853e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "hom_ref = df[df['reference']=='=']\n",
    "hom_ref = hom_ref[hom_ref['alleleSeq'] == '=']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "24e933aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "hom_ref['hom_ref_count'] = hom_ref['end'].astype(int) - hom_ref['begin'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a8593668",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&gt;locus</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>allele</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>varType</th>\n",
       "      <th>reference</th>\n",
       "      <th>alleleSeq</th>\n",
       "      <th>varScoreVAF</th>\n",
       "      <th>varScoreEAF</th>\n",
       "      <th>varFilter</th>\n",
       "      <th>hapLink</th>\n",
       "      <th>xRef</th>\n",
       "      <th>alleleFreq</th>\n",
       "      <th>alternativeCalls</th>\n",
       "      <th>hom_ref_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11430</td>\n",
       "      <td>11452</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11481</td>\n",
       "      <td>11504</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11511</td>\n",
       "      <td>11523</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11575</td>\n",
       "      <td>11583</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11609</td>\n",
       "      <td>11648</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   >locus ploidy allele chromosome  begin    end varType reference alleleSeq  \\\n",
       "3       3      2    all       chr1  11430  11452     ref         =         =   \n",
       "5       5      2    all       chr1  11481  11504     ref         =         =   \n",
       "7       7      2    all       chr1  11511  11523     ref         =         =   \n",
       "9       9      2    all       chr1  11575  11583     ref         =         =   \n",
       "11     11      2    all       chr1  11609  11648     ref         =         =   \n",
       "\n",
       "   varScoreVAF varScoreEAF varFilter hapLink xRef alleleFreq alternativeCalls  \\\n",
       "3          NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "5          NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "7          NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "9          NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "11         NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "\n",
       "    hom_ref_count  \n",
       "3              22  \n",
       "5              23  \n",
       "7              12  \n",
       "9               8  \n",
       "11             39  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hom_ref.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f9a450b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2789813668\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "141.36365310141306"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_hom_ref = hom_ref['hom_ref_count'].sum()\n",
    "print(total_hom_ref+orig_len)\n",
    "total_hom_ref/orig_len"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19815478",
   "metadata": {},
   "source": [
    "# Step 1: Create Complete chromosome 1 (Golden Reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "98774bcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr1 = pd.DataFrame()\n",
    "chr1 = pd.concat([df, chr1])\n",
    "chr1 = chr1[chr1['chromosome'] == 'chr1']\n",
    "chr1 = filter_chr(chr1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8d3996bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&gt;locus</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>allele</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>varType</th>\n",
       "      <th>reference</th>\n",
       "      <th>alleleSeq</th>\n",
       "      <th>varScoreVAF</th>\n",
       "      <th>varScoreEAF</th>\n",
       "      <th>varFilter</th>\n",
       "      <th>hapLink</th>\n",
       "      <th>xRef</th>\n",
       "      <th>alleleFreq</th>\n",
       "      <th>alternativeCalls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11430</td>\n",
       "      <td>11452</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11481</td>\n",
       "      <td>11504</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11511</td>\n",
       "      <td>11523</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11575</td>\n",
       "      <td>11583</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11609</td>\n",
       "      <td>11648</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569431</th>\n",
       "      <td>1237918</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249238327</td>\n",
       "      <td>249238343</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569433</th>\n",
       "      <td>1237920</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249238350</td>\n",
       "      <td>249238410</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569435</th>\n",
       "      <td>1237922</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249238436</td>\n",
       "      <td>249238468</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569437</th>\n",
       "      <td>1237924</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249238469</td>\n",
       "      <td>249238479</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569439</th>\n",
       "      <td>1237926</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249238556</td>\n",
       "      <td>249238568</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1148580 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          >locus ploidy allele chromosome      begin        end varType  \\\n",
       "3              3      2    all       chr1      11430      11452     ref   \n",
       "5              5      2    all       chr1      11481      11504     ref   \n",
       "7              7      2    all       chr1      11511      11523     ref   \n",
       "9              9      2    all       chr1      11575      11583     ref   \n",
       "11            11      2    all       chr1      11609      11648     ref   \n",
       "...          ...    ...    ...        ...        ...        ...     ...   \n",
       "1569431  1237918      2    all       chr1  249238327  249238343     ref   \n",
       "1569433  1237920      2    all       chr1  249238350  249238410     ref   \n",
       "1569435  1237922      2    all       chr1  249238436  249238468     ref   \n",
       "1569437  1237924      2    all       chr1  249238469  249238479     ref   \n",
       "1569439  1237926      2    all       chr1  249238556  249238568     ref   \n",
       "\n",
       "        reference alleleSeq varScoreVAF varScoreEAF varFilter hapLink xRef  \\\n",
       "3               =         =         NaN         NaN       NaN     NaN    .   \n",
       "5               =         =         NaN         NaN       NaN     NaN    .   \n",
       "7               =         =         NaN         NaN       NaN     NaN    .   \n",
       "9               =         =         NaN         NaN       NaN     NaN    .   \n",
       "11              =         =         NaN         NaN       NaN     NaN    .   \n",
       "...           ...       ...         ...         ...       ...     ...  ...   \n",
       "1569431         =         =         NaN         NaN       NaN     NaN    .   \n",
       "1569433         =         =         NaN         NaN       NaN     NaN    .   \n",
       "1569435         =         =         NaN         NaN       NaN     NaN    .   \n",
       "1569437         =         =         NaN         NaN       NaN     NaN    .   \n",
       "1569439         =         =         NaN         NaN       NaN     NaN    .   \n",
       "\n",
       "        alleleFreq alternativeCalls  \n",
       "3              NaN              NaN  \n",
       "5              NaN              NaN  \n",
       "7              NaN              NaN  \n",
       "9              NaN              NaN  \n",
       "11             NaN              NaN  \n",
       "...            ...              ...  \n",
       "1569431        NaN              NaN  \n",
       "1569433        NaN              NaN  \n",
       "1569435        NaN              NaN  \n",
       "1569437        NaN              NaN  \n",
       "1569439        NaN              NaN  \n",
       "\n",
       "[1148580 rows x 16 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "944908e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&gt;locus</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>allele</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>varType</th>\n",
       "      <th>reference</th>\n",
       "      <th>alleleSeq</th>\n",
       "      <th>varScoreVAF</th>\n",
       "      <th>varScoreEAF</th>\n",
       "      <th>varFilter</th>\n",
       "      <th>hapLink</th>\n",
       "      <th>xRef</th>\n",
       "      <th>alleleFreq</th>\n",
       "      <th>alternativeCalls</th>\n",
       "      <th>hom_ref_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11430</td>\n",
       "      <td>11452</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11481</td>\n",
       "      <td>11504</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11511</td>\n",
       "      <td>11523</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11575</td>\n",
       "      <td>11583</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11609</td>\n",
       "      <td>11648</td>\n",
       "      <td>ref</td>\n",
       "      <td>=</td>\n",
       "      <td>=</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   >locus ploidy allele chromosome  begin    end varType reference alleleSeq  \\\n",
       "3       3      2    all       chr1  11430  11452     ref         =         =   \n",
       "5       5      2    all       chr1  11481  11504     ref         =         =   \n",
       "7       7      2    all       chr1  11511  11523     ref         =         =   \n",
       "9       9      2    all       chr1  11575  11583     ref         =         =   \n",
       "11     11      2    all       chr1  11609  11648     ref         =         =   \n",
       "\n",
       "   varScoreVAF varScoreEAF varFilter hapLink xRef alleleFreq alternativeCalls  \\\n",
       "3          NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "5          NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "7          NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "9          NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "11         NaN         NaN       NaN     NaN  NaN        NaN              NaN   \n",
       "\n",
       "    hom_ref_count  \n",
       "3              22  \n",
       "5              23  \n",
       "7              12  \n",
       "9               8  \n",
       "11             39  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_hom_ref = hom_ref[hom_ref['chromosome'] == 'chr1']\n",
    "chr1_hom_ref.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "62387983",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr1_hom_ref.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ca7368c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&gt;locus</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>allele</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>varType</th>\n",
       "      <th>reference</th>\n",
       "      <th>alleleSeq</th>\n",
       "      <th>varScoreVAF</th>\n",
       "      <th>varScoreEAF</th>\n",
       "      <th>varFilter</th>\n",
       "      <th>hapLink</th>\n",
       "      <th>xRef</th>\n",
       "      <th>alleleFreq</th>\n",
       "      <th>alternativeCalls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>774</th>\n",
       "      <td>774</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38231</td>\n",
       "      <td>38232</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>105</td>\n",
       "      <td>105</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>774</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38231</td>\n",
       "      <td>38232</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>239</td>\n",
       "      <td>239</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>792</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38906</td>\n",
       "      <td>38907</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_0_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>792</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38906</td>\n",
       "      <td>38907</td>\n",
       "      <td>snp</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_0_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>;;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>852</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>41980</td>\n",
       "      <td>41981</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>482</td>\n",
       "      <td>482</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs806721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>855</th>\n",
       "      <td>852</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>41980</td>\n",
       "      <td>41981</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>179</td>\n",
       "      <td>179</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs806721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    >locus ploidy allele chromosome  begin    end varType reference alleleSeq  \\\n",
       "774    774      2      1       chr1  38231  38232     snp         A         G   \n",
       "775    774      2      2       chr1  38231  38232     snp         A         G   \n",
       "793    792      2      1       chr1  38906  38907     ref         C         C   \n",
       "794    792      2      2       chr1  38906  38907     snp         C         T   \n",
       "854    852      2      1       chr1  41980  41981     snp         A         G   \n",
       "855    852      2      2       chr1  41980  41981     snp         A         G   \n",
       "\n",
       "    varScoreVAF varScoreEAF varFilter       hapLink      xRef alleleFreq  \\\n",
       "774         105         105      PASS           NaN         .          ;   \n",
       "775         239         239      PASS           NaN         .          ;   \n",
       "793          64          64      PASS  Phased_0_0_0         .        NaN   \n",
       "794         125         125      PASS  Phased_0_0_1         .         ;;   \n",
       "854         482         482      PASS           NaN  rs806721        NaN   \n",
       "855         179         179      PASS           NaN  rs806721        NaN   \n",
       "\n",
       "    alternativeCalls  \n",
       "774              NaN  \n",
       "775              NaN  \n",
       "793              NaN  \n",
       "794              NaN  \n",
       "854              NaN  \n",
       "855              NaN  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_vars = pd.DataFrame()\n",
    "chr1_vars = pd.concat([chr1[chr1['reference'] !='='], chr1_vars])\n",
    "chr1_vars.head(6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "874f10c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1148580\n",
      "529839\n",
      "618741\n"
     ]
    }
   ],
   "source": [
    "print(len(chr1))\n",
    "print(len(chr1_vars))\n",
    "print(len(chr1_hom_ref)) #Not including the count of begin and end ranges"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de84988c",
   "metadata": {},
   "source": [
    "### The following cell iteratres through the chromosome 1 dataframe that only contains homozygous reference regions. For each homozygous reference call in the df, samtools is called to extract the corresponding region from the GRCh37 reference file and appends the reference positions to a new dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "de4e357c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9700.41015625\n",
      "7296.06640625\n",
      "0/1148580\n",
      "1000/1148580\n",
      "2000/1148580\n",
      "3000/1148580\n",
      "4000/1148580\n",
      "5000/1148580\n",
      "6000/1148580\n",
      "7000/1148580\n",
      "8000/1148580\n",
      "9000/1148580\n",
      "10000/1148580\n",
      "11000/1148580\n",
      "12000/1148580\n",
      "13000/1148580\n",
      "14000/1148580\n",
      "15000/1148580\n",
      "16000/1148580\n",
      "17000/1148580\n",
      "18000/1148580\n",
      "19000/1148580\n",
      "20000/1148580\n",
      "21000/1148580\n",
      "22000/1148580\n",
      "23000/1148580\n",
      "24000/1148580\n",
      "25000/1148580\n",
      "26000/1148580\n",
      "27000/1148580\n",
      "28000/1148580\n",
      "29000/1148580\n",
      "30000/1148580\n",
      "31000/1148580\n",
      "32000/1148580\n",
      "33000/1148580\n",
      "34000/1148580\n",
      "35000/1148580\n",
      "36000/1148580\n",
      "37000/1148580\n",
      "38000/1148580\n",
      "39000/1148580\n",
      "40000/1148580\n",
      "41000/1148580\n",
      "42000/1148580\n",
      "43000/1148580\n",
      "44000/1148580\n",
      "45000/1148580\n",
      "46000/1148580\n",
      "47000/1148580\n",
      "48000/1148580\n",
      "49000/1148580\n",
      "50000/1148580\n"
     ]
    }
   ],
   "source": [
    "sam_dir = '/Users/jerenolsen/Desktop/All_Tests/Samtools'\n",
    "cols = chr1.columns\n",
    "new_entries = {col:[] for col in cols}\n",
    "LOCUS = int(df['>locus'].astype(int).max())\n",
    "len_chr1 = len(chr1)\n",
    "\n",
    "unique_ref = 0\n",
    "for num, row in chr1_hom_ref.iterrows():\n",
    "    #t0=time.time()\n",
    "    beg = None\n",
    "    end = None\n",
    "    extact_script = None\n",
    "    if row['reference'] == '=':\n",
    "        beg = int(row['begin'])\n",
    "        end = int(row['end'])\n",
    "    else:\n",
    "        continue\n",
    "        \n",
    "    extract_script = \"samtools faidx {sam_dir}/Homo_sapiens.GRCh37.dna.primary_assembly.fa 1:{beg}-{end}\".format(beg=beg, end=end, sam_dir = sam_dir).split(\" \")\n",
    "\n",
    "    process = subprocess.run(extract_script, capture_output=True, text=True)\n",
    "    outs = process.stdout.split()[1:]\n",
    "    outs = \"\".join(outs)\n",
    "    \n",
    "    genotypes = list(outs)\n",
    "    genotypes = list(\"\".join(genotypes))\n",
    "    num_positions = end-beg\n",
    "    \n",
    "    unique_ref+=1\n",
    "    \n",
    "    for i in range(0,num_positions):\n",
    "        LOCUS +=1\n",
    "        for j in range(1,3):\n",
    "            genotype = genotypes[i]\n",
    "            \n",
    "            new_entries['>locus'].append(LOCUS)\n",
    "            new_entries['ploidy'].append(2)\n",
    "            new_entries['allele'].append(j)\n",
    "            new_entries['chromosome'].append('chr1')\n",
    "            new_entries['begin'].append(beg+i)\n",
    "            new_entries['end'].append(beg+i+1)\n",
    "            new_entries['varType'].append('ref')\n",
    "            new_entries['reference'].append(genotype)\n",
    "            new_entries['alleleSeq'].append(genotype)\n",
    "            new_entries['varScoreVAF'].append(np.nan)\n",
    "            new_entries['varScoreEAF'].append(np.nan)\n",
    "            new_entries['varFilter'].append(np.nan)\n",
    "            new_entries['hapLink'].append(np.nan)\n",
    "            new_entries['xRef'].append(np.nan)\n",
    "            new_entries['alleleFreq'].append(np.nan)\n",
    "            new_entries['alternativeCalls'].append(np.nan)\n",
    "\n",
    "    if num%1000 ==0:\n",
    "        print(f\"{num}/{len_chr1}\")\n",
    "        \n",
    "    if num == 50000:\n",
    "        break\n",
    "        \n",
    "df_hom_ref_chr1 = pd.DataFrame.from_dict(new_entries)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "090853d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16686438"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_hom_ref_chr1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9df517de",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get all variant positions from orignal chr1 dataframe that go up to the max position extracted from reference file\n",
    "# in df_hom_ref_chr1\n",
    "max_begin = df_hom_ref_chr1['begin'].astype(int).max()\n",
    "chr1_variants = chr1_vars[chr1_vars['begin'].astype(int) <max_begin]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e4412818",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&gt;locus</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>allele</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>varType</th>\n",
       "      <th>reference</th>\n",
       "      <th>alleleSeq</th>\n",
       "      <th>varScoreVAF</th>\n",
       "      <th>varScoreEAF</th>\n",
       "      <th>varFilter</th>\n",
       "      <th>hapLink</th>\n",
       "      <th>xRef</th>\n",
       "      <th>alleleFreq</th>\n",
       "      <th>alternativeCalls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>774</th>\n",
       "      <td>774</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38231</td>\n",
       "      <td>38232</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>105</td>\n",
       "      <td>105</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>774</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38231</td>\n",
       "      <td>38232</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>239</td>\n",
       "      <td>239</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>792</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38906</td>\n",
       "      <td>38907</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_0_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>792</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38906</td>\n",
       "      <td>38907</td>\n",
       "      <td>snp</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_0_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>;;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>852</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>41980</td>\n",
       "      <td>41981</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>482</td>\n",
       "      <td>482</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs806721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114610</th>\n",
       "      <td>99839</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9670833</td>\n",
       "      <td>9670834</td>\n",
       "      <td>snp</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>247</td>\n",
       "      <td>247</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_5_71_1</td>\n",
       "      <td>rs4926472</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114636</th>\n",
       "      <td>99865</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9672090</td>\n",
       "      <td>9672091</td>\n",
       "      <td>snp</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>361</td>\n",
       "      <td>361</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs4073293</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114637</th>\n",
       "      <td>99865</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9672090</td>\n",
       "      <td>9672091</td>\n",
       "      <td>snp</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs4073293</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114767</th>\n",
       "      <td>99993</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9675375</td>\n",
       "      <td>9675376</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_5_71_1</td>\n",
       "      <td>rs4078500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114768</th>\n",
       "      <td>99993</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9675375</td>\n",
       "      <td>9675376</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>153</td>\n",
       "      <td>153</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_5_71_0</td>\n",
       "      <td>rs4078500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21375 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       >locus ploidy allele chromosome    begin      end varType reference  \\\n",
       "774       774      2      1       chr1    38231    38232     snp         A   \n",
       "775       774      2      2       chr1    38231    38232     snp         A   \n",
       "793       792      2      1       chr1    38906    38907     ref         C   \n",
       "794       792      2      2       chr1    38906    38907     snp         C   \n",
       "854       852      2      1       chr1    41980    41981     snp         A   \n",
       "...       ...    ...    ...        ...      ...      ...     ...       ...   \n",
       "114610  99839      2      2       chr1  9670833  9670834     snp         C   \n",
       "114636  99865      2      1       chr1  9672090  9672091     snp         G   \n",
       "114637  99865      2      2       chr1  9672090  9672091     snp         G   \n",
       "114767  99993      2      1       chr1  9675375  9675376     snp         A   \n",
       "114768  99993      2      2       chr1  9675375  9675376     snp         A   \n",
       "\n",
       "       alleleSeq varScoreVAF varScoreEAF varFilter        hapLink       xRef  \\\n",
       "774            G         105         105      PASS            NaN          .   \n",
       "775            G         239         239      PASS            NaN          .   \n",
       "793            C          64          64      PASS   Phased_0_0_0          .   \n",
       "794            T         125         125      PASS   Phased_0_0_1          .   \n",
       "854            G         482         482      PASS            NaN   rs806721   \n",
       "...          ...         ...         ...       ...            ...        ...   \n",
       "114610         T         247         247      PASS  Phased_5_71_1  rs4926472   \n",
       "114636         A         361         361      PASS            NaN  rs4073293   \n",
       "114637         A          43          43      PASS            NaN  rs4073293   \n",
       "114767         G          34          34      PASS  Phased_5_71_1  rs4078500   \n",
       "114768         G         153         153      PASS  Phased_5_71_0  rs4078500   \n",
       "\n",
       "       alleleFreq alternativeCalls  \n",
       "774             ;              NaN  \n",
       "775             ;              NaN  \n",
       "793           NaN              NaN  \n",
       "794            ;;              NaN  \n",
       "854           NaN              NaN  \n",
       "...           ...              ...  \n",
       "114610        NaN              NaN  \n",
       "114636        NaN              NaN  \n",
       "114637        NaN              NaN  \n",
       "114767        NaN              NaN  \n",
       "114768        NaN              NaN  \n",
       "\n",
       "[21375 rows x 16 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c774b737",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21375"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(chr1_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e7a111fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove the reference ranges in orignal chr1 df \n",
    "chr1_variants = chr1_variant[chr1_variants['reference'] != '=']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "8a77abc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&gt;locus</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>allele</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>varType</th>\n",
       "      <th>reference</th>\n",
       "      <th>alleleSeq</th>\n",
       "      <th>varScoreVAF</th>\n",
       "      <th>varScoreEAF</th>\n",
       "      <th>varFilter</th>\n",
       "      <th>hapLink</th>\n",
       "      <th>xRef</th>\n",
       "      <th>alleleFreq</th>\n",
       "      <th>alternativeCalls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>774</th>\n",
       "      <td>774</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38231</td>\n",
       "      <td>38232</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>105</td>\n",
       "      <td>105</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>774</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38231</td>\n",
       "      <td>38232</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>239</td>\n",
       "      <td>239</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>792</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38906</td>\n",
       "      <td>38907</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_0_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>792</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38906</td>\n",
       "      <td>38907</td>\n",
       "      <td>snp</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_0_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>;;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>852</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>41980</td>\n",
       "      <td>41981</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>482</td>\n",
       "      <td>482</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs806721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114391</th>\n",
       "      <td>99633</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9653834</td>\n",
       "      <td>9653835</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>242</td>\n",
       "      <td>242</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs3923984</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114392</th>\n",
       "      <td>99633</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9653834</td>\n",
       "      <td>9653835</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs3923984</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114396</th>\n",
       "      <td>99637</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9654196</td>\n",
       "      <td>9654197</td>\n",
       "      <td>snp</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>467</td>\n",
       "      <td>467</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs7544083</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114397</th>\n",
       "      <td>99637</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9654196</td>\n",
       "      <td>9654197</td>\n",
       "      <td>snp</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs7544083</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114425</th>\n",
       "      <td>99665</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9656570</td>\n",
       "      <td>9656571</td>\n",
       "      <td>snp</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_5_71_1</td>\n",
       "      <td>rs4074744</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21355 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       >locus ploidy allele chromosome    begin      end varType reference  \\\n",
       "774       774      2      1       chr1    38231    38232     snp         A   \n",
       "775       774      2      2       chr1    38231    38232     snp         A   \n",
       "793       792      2      1       chr1    38906    38907     ref         C   \n",
       "794       792      2      2       chr1    38906    38907     snp         C   \n",
       "854       852      2      1       chr1    41980    41981     snp         A   \n",
       "...       ...    ...    ...        ...      ...      ...     ...       ...   \n",
       "114391  99633      2      1       chr1  9653834  9653835     snp         A   \n",
       "114392  99633      2      2       chr1  9653834  9653835     snp         A   \n",
       "114396  99637      2      1       chr1  9654196  9654197     snp         T   \n",
       "114397  99637      2      2       chr1  9654196  9654197     snp         T   \n",
       "114425  99665      2      1       chr1  9656570  9656571     snp         G   \n",
       "\n",
       "       alleleSeq varScoreVAF varScoreEAF varFilter        hapLink       xRef  \\\n",
       "774            G         105         105      PASS            NaN          .   \n",
       "775            G         239         239      PASS            NaN          .   \n",
       "793            C          64          64      PASS   Phased_0_0_0          .   \n",
       "794            T         125         125      PASS   Phased_0_0_1          .   \n",
       "854            G         482         482      PASS            NaN   rs806721   \n",
       "...          ...         ...         ...       ...            ...        ...   \n",
       "114391         G         242         242      PASS            NaN  rs3923984   \n",
       "114392         G          25          25      PASS            NaN  rs3923984   \n",
       "114396         C         467         467      PASS            NaN  rs7544083   \n",
       "114397         C          54          54      PASS            NaN  rs7544083   \n",
       "114425         T          46          46      PASS  Phased_5_71_1  rs4074744   \n",
       "\n",
       "       alleleFreq alternativeCalls  \n",
       "774             ;              NaN  \n",
       "775             ;              NaN  \n",
       "793           NaN              NaN  \n",
       "794            ;;              NaN  \n",
       "854           NaN              NaN  \n",
       "...           ...              ...  \n",
       "114391        NaN              NaN  \n",
       "114392        NaN              NaN  \n",
       "114396        NaN              NaN  \n",
       "114397        NaN              NaN  \n",
       "114425        NaN              NaN  \n",
       "\n",
       "[21355 rows x 16 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_variants.head(-20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6713013",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Combine chr1 variant calls and reference calls back together for complete chromosome subrange"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6c91fe32",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr1_combined = pd.concat([chr1_variants, df_hom_ref_chr1 ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "577d6a14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&gt;locus</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>allele</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>varType</th>\n",
       "      <th>reference</th>\n",
       "      <th>alleleSeq</th>\n",
       "      <th>varScoreVAF</th>\n",
       "      <th>varScoreEAF</th>\n",
       "      <th>varFilter</th>\n",
       "      <th>hapLink</th>\n",
       "      <th>xRef</th>\n",
       "      <th>alleleFreq</th>\n",
       "      <th>alternativeCalls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>774</th>\n",
       "      <td>774</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38231</td>\n",
       "      <td>38232</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>105</td>\n",
       "      <td>105</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>774</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38231</td>\n",
       "      <td>38232</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>239</td>\n",
       "      <td>239</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>792</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38906</td>\n",
       "      <td>38907</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_0_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>792</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38906</td>\n",
       "      <td>38907</td>\n",
       "      <td>snp</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Phased_0_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>;;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>852</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>41980</td>\n",
       "      <td>41981</td>\n",
       "      <td>snp</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>482</td>\n",
       "      <td>482</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs806721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686433</th>\n",
       "      <td>23766500</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676811</td>\n",
       "      <td>9676812</td>\n",
       "      <td>ref</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686434</th>\n",
       "      <td>23766501</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676812</td>\n",
       "      <td>9676813</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686435</th>\n",
       "      <td>23766501</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676812</td>\n",
       "      <td>9676813</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686436</th>\n",
       "      <td>23766502</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676813</td>\n",
       "      <td>9676814</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686437</th>\n",
       "      <td>23766502</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676813</td>\n",
       "      <td>9676814</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16707813 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            >locus ploidy allele chromosome    begin      end varType  \\\n",
       "774            774      2      1       chr1    38231    38232     snp   \n",
       "775            774      2      2       chr1    38231    38232     snp   \n",
       "793            792      2      1       chr1    38906    38907     ref   \n",
       "794            792      2      2       chr1    38906    38907     snp   \n",
       "854            852      2      1       chr1    41980    41981     snp   \n",
       "...            ...    ...    ...        ...      ...      ...     ...   \n",
       "16686433  23766500      2      2       chr1  9676811  9676812     ref   \n",
       "16686434  23766501      2      1       chr1  9676812  9676813     ref   \n",
       "16686435  23766501      2      2       chr1  9676812  9676813     ref   \n",
       "16686436  23766502      2      1       chr1  9676813  9676814     ref   \n",
       "16686437  23766502      2      2       chr1  9676813  9676814     ref   \n",
       "\n",
       "         reference alleleSeq varScoreVAF varScoreEAF varFilter       hapLink  \\\n",
       "774              A         G         105         105      PASS           NaN   \n",
       "775              A         G         239         239      PASS           NaN   \n",
       "793              C         C          64          64      PASS  Phased_0_0_0   \n",
       "794              C         T         125         125      PASS  Phased_0_0_1   \n",
       "854              A         G         482         482      PASS           NaN   \n",
       "...            ...       ...         ...         ...       ...           ...   \n",
       "16686433         G         G         NaN         NaN       NaN           NaN   \n",
       "16686434         C         C         NaN         NaN       NaN           NaN   \n",
       "16686435         C         C         NaN         NaN       NaN           NaN   \n",
       "16686436         C         C         NaN         NaN       NaN           NaN   \n",
       "16686437         C         C         NaN         NaN       NaN           NaN   \n",
       "\n",
       "              xRef alleleFreq alternativeCalls  \n",
       "774              .          ;              NaN  \n",
       "775              .          ;              NaN  \n",
       "793              .        NaN              NaN  \n",
       "794              .         ;;              NaN  \n",
       "854       rs806721        NaN              NaN  \n",
       "...            ...        ...              ...  \n",
       "16686433       NaN        NaN              NaN  \n",
       "16686434       NaN        NaN              NaN  \n",
       "16686435       NaN        NaN              NaN  \n",
       "16686436       NaN        NaN              NaN  \n",
       "16686437       NaN        NaN              NaN  \n",
       "\n",
       "[16707813 rows x 16 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "853b57b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&gt;locus</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>allele</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>varType</th>\n",
       "      <th>reference</th>\n",
       "      <th>alleleSeq</th>\n",
       "      <th>varScoreVAF</th>\n",
       "      <th>varScoreEAF</th>\n",
       "      <th>varFilter</th>\n",
       "      <th>hapLink</th>\n",
       "      <th>xRef</th>\n",
       "      <th>alleleFreq</th>\n",
       "      <th>alternativeCalls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15423284</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11430</td>\n",
       "      <td>11431</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15423284</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11430</td>\n",
       "      <td>11431</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15423285</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11431</td>\n",
       "      <td>11432</td>\n",
       "      <td>ref</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15423285</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11431</td>\n",
       "      <td>11432</td>\n",
       "      <td>ref</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15423286</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11432</td>\n",
       "      <td>11433</td>\n",
       "      <td>ref</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686433</th>\n",
       "      <td>23766500</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676811</td>\n",
       "      <td>9676812</td>\n",
       "      <td>ref</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686434</th>\n",
       "      <td>23766501</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676812</td>\n",
       "      <td>9676813</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686435</th>\n",
       "      <td>23766501</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676812</td>\n",
       "      <td>9676813</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686436</th>\n",
       "      <td>23766502</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676813</td>\n",
       "      <td>9676814</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16686437</th>\n",
       "      <td>23766502</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9676813</td>\n",
       "      <td>9676814</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16707813 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            >locus ploidy allele chromosome    begin      end varType  \\\n",
       "0         15423284      2      1       chr1    11430    11431     ref   \n",
       "1         15423284      2      2       chr1    11430    11431     ref   \n",
       "2         15423285      2      1       chr1    11431    11432     ref   \n",
       "3         15423285      2      2       chr1    11431    11432     ref   \n",
       "4         15423286      2      1       chr1    11432    11433     ref   \n",
       "...            ...    ...    ...        ...      ...      ...     ...   \n",
       "16686433  23766500      2      2       chr1  9676811  9676812     ref   \n",
       "16686434  23766501      2      1       chr1  9676812  9676813     ref   \n",
       "16686435  23766501      2      2       chr1  9676812  9676813     ref   \n",
       "16686436  23766502      2      1       chr1  9676813  9676814     ref   \n",
       "16686437  23766502      2      2       chr1  9676813  9676814     ref   \n",
       "\n",
       "         reference alleleSeq varScoreVAF varScoreEAF varFilter hapLink xRef  \\\n",
       "0                C         C         NaN         NaN       NaN     NaN  NaN   \n",
       "1                C         C         NaN         NaN       NaN     NaN  NaN   \n",
       "2                T         T         NaN         NaN       NaN     NaN  NaN   \n",
       "3                T         T         NaN         NaN       NaN     NaN  NaN   \n",
       "4                G         G         NaN         NaN       NaN     NaN  NaN   \n",
       "...            ...       ...         ...         ...       ...     ...  ...   \n",
       "16686433         G         G         NaN         NaN       NaN     NaN  NaN   \n",
       "16686434         C         C         NaN         NaN       NaN     NaN  NaN   \n",
       "16686435         C         C         NaN         NaN       NaN     NaN  NaN   \n",
       "16686436         C         C         NaN         NaN       NaN     NaN  NaN   \n",
       "16686437         C         C         NaN         NaN       NaN     NaN  NaN   \n",
       "\n",
       "         alleleFreq alternativeCalls  \n",
       "0               NaN              NaN  \n",
       "1               NaN              NaN  \n",
       "2               NaN              NaN  \n",
       "3               NaN              NaN  \n",
       "4               NaN              NaN  \n",
       "...             ...              ...  \n",
       "16686433        NaN              NaN  \n",
       "16686434        NaN              NaN  \n",
       "16686435        NaN              NaN  \n",
       "16686436        NaN              NaN  \n",
       "16686437        NaN              NaN  \n",
       "\n",
       "[16707813 rows x 16 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_combined['begin'] = chr1_combined['begin'].astype(int)\n",
    "chr1_combined.sort_values(by=['begin','>locus'], inplace = True)\n",
    "chr1_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "0e1895f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16707813"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(chr1_combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "925fd2db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&gt;locus</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>allele</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>varType</th>\n",
       "      <th>reference</th>\n",
       "      <th>alleleSeq</th>\n",
       "      <th>varScoreVAF</th>\n",
       "      <th>varScoreEAF</th>\n",
       "      <th>varFilter</th>\n",
       "      <th>hapLink</th>\n",
       "      <th>xRef</th>\n",
       "      <th>alleleFreq</th>\n",
       "      <th>alternativeCalls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15423284</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11430</td>\n",
       "      <td>11431</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15423284</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11430</td>\n",
       "      <td>11431</td>\n",
       "      <td>ref</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15423285</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11431</td>\n",
       "      <td>11432</td>\n",
       "      <td>ref</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15423285</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11431</td>\n",
       "      <td>11432</td>\n",
       "      <td>ref</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15423286</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11432</td>\n",
       "      <td>11433</td>\n",
       "      <td>ref</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     >locus ploidy allele chromosome  begin    end varType reference  \\\n",
       "0  15423284      2      1       chr1  11430  11431     ref         C   \n",
       "1  15423284      2      2       chr1  11430  11431     ref         C   \n",
       "2  15423285      2      1       chr1  11431  11432     ref         T   \n",
       "3  15423285      2      2       chr1  11431  11432     ref         T   \n",
       "4  15423286      2      1       chr1  11432  11433     ref         G   \n",
       "\n",
       "  alleleSeq varScoreVAF varScoreEAF varFilter hapLink xRef alleleFreq  \\\n",
       "0         C         NaN         NaN       NaN     NaN  NaN        NaN   \n",
       "1         C         NaN         NaN       NaN     NaN  NaN        NaN   \n",
       "2         T         NaN         NaN       NaN     NaN  NaN        NaN   \n",
       "3         T         NaN         NaN       NaN     NaN  NaN        NaN   \n",
       "4         G         NaN         NaN       NaN     NaN  NaN        NaN   \n",
       "\n",
       "  alternativeCalls  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              NaN  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame()\n",
    "df = pd.concat([chr1_combined,df])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be1f2a45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to VCF format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a237bd98",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/7s/_bh70s2n68zd14b4fsfcjcv00000gn/T/ipykernel_903/55039304.py:24: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['orig_ref_alt'] = df['REF'] + df['ALT']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           >locus ploidy allele CHROM    begin      POS varType REF ALT  \\\n",
      "1387379  16116427      2      1  chr1  1212635  1212636     ref   G   G   \n",
      "1387380  16116427      2      2  chr1  1212635  1212636     ref   G   G   \n",
      "1387377  16116426      2      1  chr1  1212634  1212635     ref   G   G   \n",
      "1387378  16116426      2      2  chr1  1212634  1212635     ref   G   G   \n",
      "1387375  16116425      2      1  chr1  1212633  1212634     ref   T   T   \n",
      "1387376  16116425      2      2  chr1  1212633  1212634     ref   T   T   \n",
      "1387373  16116424      2      1  chr1  1212632  1212633     ref   C   C   \n",
      "1387374  16116424      2      2  chr1  1212632  1212633     ref   C   C   \n",
      "1387371  16116423      2      1  chr1  1212631  1212632     ref   G   G   \n",
      "1387372  16116423      2      2  chr1  1212631  1212632     ref   G   G   \n",
      "1387369  16116422      2      1  chr1  1212630  1212631     ref   A   A   \n",
      "1387370  16116422      2      2  chr1  1212630  1212631     ref   A   A   \n",
      "1387367  16116421      2      1  chr1  1212629  1212630     ref   T   T   \n",
      "1387368  16116421      2      2  chr1  1212629  1212630     ref   T   T   \n",
      "1387365  16116420      2      1  chr1  1212628  1212629     ref   G   G   \n",
      "1387366  16116420      2      2  chr1  1212628  1212629     ref   G   G   \n",
      "1387363  16116419      2      1  chr1  1212627  1212628     ref   A   A   \n",
      "1387364  16116419      2      2  chr1  1212627  1212628     ref   A   A   \n",
      "1387361  16116418      2      1  chr1  1212626  1212627     ref   A   A   \n",
      "1387362  16116418      2      2  chr1  1212626  1212627     ref   A   A   \n",
      "1387359  16116417      2      1  chr1  1212624  1212625     ref   C   C   \n",
      "1387360  16116417      2      2  chr1  1212624  1212625     ref   C   C   \n",
      "1387357  16116416      2      1  chr1  1212623  1212624     ref   C   C   \n",
      "1387358  16116416      2      2  chr1  1212623  1212624     ref   C   C   \n",
      "1387355  16116415      2      1  chr1  1212622  1212623     ref   T   T   \n",
      "1387356  16116415      2      2  chr1  1212622  1212623     ref   T   T   \n",
      "1387353  16116414      2      1  chr1  1212621  1212622     ref   T   T   \n",
      "1387354  16116414      2      2  chr1  1212621  1212622     ref   T   T   \n",
      "1387351  16116413      2      1  chr1  1212617  1212618     ref   C   C   \n",
      "1387352  16116413      2      2  chr1  1212617  1212618     ref   C   C   \n",
      "1387349  16116412      2      1  chr1  1212616  1212617     ref   T   T   \n",
      "1387350  16116412      2      2  chr1  1212616  1212617     ref   T   T   \n",
      "1387347  16116411      2      1  chr1  1212615  1212616     ref   C   C   \n",
      "1387348  16116411      2      2  chr1  1212615  1212616     ref   C   C   \n",
      "\n",
      "        varScoreVAF  ...   ID alleleFreq alternativeCalls QUAL FILTER INFO  \\\n",
      "1387379         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387380         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387377         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387378         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387375         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387376         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387373         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387374         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387371         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387372         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387369         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387370         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387367         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387368         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387365         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387366         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387363         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387364         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387361         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387362         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387359         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387360         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387357         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387358         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387355         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387356         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387353         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387354         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387351         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387352         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387349         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387350         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387347         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "1387348         NaN  ...  NaN        NaN              NaN    .      .    .   \n",
      "\n",
      "        FORMAT orig_ref_alt NA00001 2nd_sort_col  \n",
      "1387379     GT           GG     0/0            1  \n",
      "1387380     GT           GG     0/0            1  \n",
      "1387377     GT           GG     0/0            1  \n",
      "1387378     GT           GG     0/0            1  \n",
      "1387375     GT           TT     0/0            1  \n",
      "1387376     GT           TT     0/0            1  \n",
      "1387373     GT           CC     0/0            1  \n",
      "1387374     GT           CC     0/0            1  \n",
      "1387371     GT           GG     0/0            1  \n",
      "1387372     GT           GG     0/0            1  \n",
      "1387369     GT           AA     0/0            1  \n",
      "1387370     GT           AA     0/0            1  \n",
      "1387367     GT           TT     0/0            1  \n",
      "1387368     GT           TT     0/0            1  \n",
      "1387365     GT           GG     0/0            1  \n",
      "1387366     GT           GG     0/0            1  \n",
      "1387363     GT           AA     0/0            1  \n",
      "1387364     GT           AA     0/0            1  \n",
      "1387361     GT           AA     0/0            1  \n",
      "1387362     GT           AA     0/0            1  \n",
      "1387359     GT           CC     0/0            1  \n",
      "1387360     GT           CC     0/0            1  \n",
      "1387357     GT           CC     0/0            1  \n",
      "1387358     GT           CC     0/0            1  \n",
      "1387355     GT           TT     0/0            1  \n",
      "1387356     GT           TT     0/0            1  \n",
      "1387353     GT           TT     0/0            1  \n",
      "1387354     GT           TT     0/0            1  \n",
      "1387351     GT           CC     0/0            1  \n",
      "1387352     GT           CC     0/0            1  \n",
      "1387349     GT           TT     0/0            1  \n",
      "1387350     GT           TT     0/0            1  \n",
      "1387347     GT           CC     0/0            1  \n",
      "1387348     GT           CC     0/0            1  \n",
      "\n",
      "[34 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "df_vcf_combined = build_vcf_df(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "cdbbc3fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NA00001</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11431</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>11432</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>11433</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>11434</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>11435</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>11436</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>11437</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>11438</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>11439</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>11440</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>11441</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>11442</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>11443</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>11444</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>11445</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>11446</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>11447</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>11448</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>11449</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>11450</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM    POS ID REF ALT QUAL FILTER INFO FORMAT NA00001\n",
       "0      1  11431  .   C   C    .      .    .     GT     0/0\n",
       "1      1  11432  .   T   T    .      .    .     GT     0/0\n",
       "2      1  11433  .   G   G    .      .    .     GT     0/0\n",
       "3      1  11434  .   C   C    .      .    .     GT     0/0\n",
       "4      1  11435  .   C   C    .      .    .     GT     0/0\n",
       "5      1  11436  .   G   G    .      .    .     GT     0/0\n",
       "6      1  11437  .   G   G    .      .    .     GT     0/0\n",
       "7      1  11438  .   G   G    .      .    .     GT     0/0\n",
       "8      1  11439  .   C   C    .      .    .     GT     0/0\n",
       "9      1  11440  .   C   C    .      .    .     GT     0/0\n",
       "10     1  11441  .   C   C    .      .    .     GT     0/0\n",
       "11     1  11442  .   T   T    .      .    .     GT     0/0\n",
       "12     1  11443  .   C   C    .      .    .     GT     0/0\n",
       "13     1  11444  .   T   T    .      .    .     GT     0/0\n",
       "14     1  11445  .   T   T    .      .    .     GT     0/0\n",
       "15     1  11446  .   G   G    .      .    .     GT     0/0\n",
       "16     1  11447  .   C   C    .      .    .     GT     0/0\n",
       "17     1  11448  .   T   T    .      .    .     GT     0/0\n",
       "18     1  11449  .   C   C    .      .    .     GT     0/0\n",
       "19     1  11450  .   C   C    .      .    .     GT     0/0"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vcf_combined.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "331ef4b4",
   "metadata": {},
   "source": [
    "# Save Golden Reference "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "fd7117ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_f = '/Users/jerenolsen/Desktop/Genomics_Pipeline/input_genomes/hu1AF4ED_CGI_RawData.tsv'\n",
    "output_f = '/Users/jerenolsen/desktop/chr1_subsection_experiment_golden_reference.vcf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2a02ead0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Build 37\n"
     ]
    }
   ],
   "source": [
    "save_file(df_vcf_combined,input_f, output_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34624b34",
   "metadata": {},
   "source": [
    "# Save Chromosome 1 subsection variants that will be input for the imputation program"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "64b79974",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/7s/_bh70s2n68zd14b4fsfcjcv00000gn/T/ipykernel_903/55039304.py:24: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['orig_ref_alt'] = df['REF'] + df['ALT']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [>locus, ploidy, allele, CHROM, begin, POS, varType, REF, ALT, varScoreVAF, varScoreEAF, varFilter, hapLink, ID, alleleFreq, alternativeCalls, QUAL, FILTER, INFO, FORMAT, orig_ref_alt, NA00001, 2nd_sort_col]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "df_vcf_vars = pd.DataFrame()\n",
    "df_vcf_vars = pd.concat([chr1_variants, df_vcf_vars])\n",
    "df_vcf_vars = build_vcf_df(df_vcf_vars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "25cc6030",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10182"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_vcf_vars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a726d888",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_vcf_vars_test = df_vcf_vars[df_vcf_vars['ALT']!= df_vcf_vars['REF']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "e56401a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10181"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_vcf_vars_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "02ae6891",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NA00001</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>38232</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>38907</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>41981</td>\n",
       "      <td>rs806721</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>47108</td>\n",
       "      <td>rs2531241</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>47292</td>\n",
       "      <td>rs2691275</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10177</th>\n",
       "      <td>1</td>\n",
       "      <td>9664650</td>\n",
       "      <td>rs138697503</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10178</th>\n",
       "      <td>1</td>\n",
       "      <td>9666364</td>\n",
       "      <td>rs7364899</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10179</th>\n",
       "      <td>1</td>\n",
       "      <td>9670483</td>\n",
       "      <td>rs4244632</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10180</th>\n",
       "      <td>1</td>\n",
       "      <td>9672091</td>\n",
       "      <td>rs4073293</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10181</th>\n",
       "      <td>1</td>\n",
       "      <td>9675376</td>\n",
       "      <td>rs4078500</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10182 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM      POS           ID REF ALT QUAL FILTER INFO FORMAT NA00001\n",
       "0         1    38232            .   A   G    .      .    .     GT     1/1\n",
       "1         1    38907            .   C   T    .      .    .     GT     0/1\n",
       "2         1    41981     rs806721   A   G    .      .    .     GT     1/1\n",
       "3         1    47108    rs2531241   G   C    .      .    .     GT     0/1\n",
       "4         1    47292    rs2691275   T   G    .      .    .     GT     0/1\n",
       "...     ...      ...          ...  ..  ..  ...    ...  ...    ...     ...\n",
       "10177     1  9664650  rs138697503   T   A    .      .    .     GT     0/1\n",
       "10178     1  9666364    rs7364899   A   G    .      .    .     GT     1/1\n",
       "10179     1  9670483    rs4244632   T   A    .      .    .     GT     1/1\n",
       "10180     1  9672091    rs4073293   G   A    .      .    .     GT     1/1\n",
       "10181     1  9675376    rs4078500   A   G    .      .    .     GT     1/1\n",
       "\n",
       "[10182 rows x 10 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vcf_vars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "efc4fba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_f = '/Users/jerenolsen/desktop/chr1_subsection_experiment_variants_tbi.vcf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "ab17cea8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Build 37\n"
     ]
    }
   ],
   "source": [
    "save_file(df_vcf_vars,input_f, output_f)\n",
    "#df_vcf_vars.to_csv('/Users/jerenolsen/desktop/chr1_subsection_experiment_variants_tbi.vcf', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f27f4d8",
   "metadata": {},
   "source": [
    "### ==============================================================\n",
    "### ===| Perform Imputation of region, experiment continues in notebook 2 | ====\n",
    "### =============================================================="
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
